File: node.js/src/slimmed/application/books/bookCommand.ts
import { PositiveNumber } from '#core/typing';
import { AuthorIdOrData } from '../../domain/books/authors';
import {
  BookId,
  ChapterContent,
  ChapterTitle,
  CommitteeApproval,
  Format,
  Genre,
  ISBN,
  PublisherId,
  Reviewer,
  Title,
  Translation,
} from '../../domain/books/entities';
import { Command } from '../../infrastructure/commands';

export type CreateDraft = Command<
  'CreateDraftCommand',
  {
    bookId: BookId;
    title: Title;
    author: AuthorIdOrData;
    publisherId: PublisherId;
    edition: PositiveNumber;
    genre: Genre | null;
  }
>;

export type AddChapter = Command<
  'AddChapterCommand',
  {
    bookId: BookId;
    chapterTitle: ChapterTitle;
    chapterContent: ChapterContent;
  }
>;

export type MoveToEditing = Command<
  'MoveToEditingCommand',
  {
    bookId: BookId;
  }
>;

export type AddTranslation = Command<
  'AddTranslationCommand',
  {
    bookId: BookId;
    translation: Translation;
  }
>;

export type AddFormat = Command<
  'AddFormatCommand',
  {
    bookId: BookId;
    format: Format;
  }
>;

export type RemoveFormat = Command<
  'RemoveFormatCommand',
  {
    bookId: BookId;
    format: Format;
  }
>;

export type AddReviewer = Command<
  'AddReviewerCommand',
  {
    bookId: BookId;
    reviewer: Reviewer;
  }
>;

export type Approve = Command<
  'ApproveCommand',
  {
    bookId: BookId;
    committeeApproval: CommitteeApproval;
  }
>;

export type SetISBN = Command<
  'SetISBNCommand',
  {
    bookId: BookId;
    isbn: ISBN;
  }
>;

export type MoveToPrinting = Command<
  'MoveToPrintingCommand',
  {
    bookId: BookId;
  }
>;

export type MoveToPublished = Command<
  'MoveToPublishedCommand',
  {
    bookId: BookId;
  }
>;

export type MoveToOutOfPrint = Command<
  'MoveToOutOfPrintCommand',
  {
    bookId: BookId;
  }
>;

export type BookCommand =
  | CreateDraft
  | AddChapter
  | MoveToEditing
  | AddTranslation
  | AddFormat
  | RemoveFormat
  | AddReviewer
  | Approve
  | SetISBN
  | MoveToPrinting
  | MoveToPublished
  | MoveToOutOfPrint;

File: csharp/Slimmed/PublishingHouse.Application/Books/BookCommand.cs
using PublishingHouse.Books.Authors;
using PublishingHouse.Books.Entities;
using PublishingHouse.Core.ValueObjects;

namespace PublishingHouse.Application.Books;

public abstract record BookCommand
{
    public record CreateDraft(
        BookId BookId,
        Title Title,
        AuthorIdOrData Author,
        PublisherId PublisherId,
        PositiveInt Edition,
        Genre? Genre
    ): BookCommand;

    public record AddChapter(BookId BookId, ChapterTitle Title, ChapterContent Content): BookCommand;

    public record MoveToEditing(BookId BookId): BookCommand;

    public record AddTranslation(BookId BookId, Translation Translation): BookCommand;

    public record AddFormat(BookId BookId, Format Format): BookCommand;

    public record RemoveFormat(BookId BookId, Format Format): BookCommand;

    public record AddReviewer(BookId BookId, Reviewer Reviewer): BookCommand;

    public record Approve(BookId BookId, CommitteeApproval CommitteeApproval): BookCommand;

    public record SetISBN(BookId BookId, ISBN ISBN): BookCommand;

    public record MoveToPrinting(BookId BookId): BookCommand;

    public record MoveToPublished(BookId BookId): BookCommand;

    public record MoveToOutOfPrint(BookId BookId): BookCommand;

    private BookCommand() { }
}
